presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CombineApproxPercentileFunctions.java

```diff
@@ -0,0 +1,278 @@
+                return 1;
+            }
+            else if (argumentTypes.size() == 4
+                    || (argumentTypes.size() == 3 && argumentTypes.get(1).getBase().equals(StandardTypes.BIGINT))) {
+                return 2;
+            }
+        }
+
+        checkState(false);
+        return -1;
+    }
+
+    private static List<RowExpression> addPercentileArgument(List<RowExpression> argumentsWithoutPercentile, RowExpression percentileArgument, int percentilePosition)
+    {
+        List<RowExpression> newAggCallArguments = new ArrayList<>();
+        newAggCallArguments.addAll(argumentsWithoutPercentile);
+        newAggCallArguments.add(percentilePosition, percentileArgument);
+        return newAggCallArguments;
+    }
+
+    // Aggregations which are the same after removing percentile arguments are candidates to be combined
+    private static AggregationNode.Aggregation aggregationWithoutPercentile(AggregationNode.Aggregation aggregation, int percentilePosition)
+    {
+        List<RowExpression> arguments = aggregation.getCall().getArguments();
+        ImmutableList.Builder<RowExpression> argumentWithoutPercentile = ImmutableList.builder();
+        for (int i = 0; i < arguments.size(); ++i) {
+            if (i != percentilePosition) {
+                argumentWithoutPercentile.add(arguments.get(i));
+            }
+        }
+
+        return new AggregationNode.Aggregation(
+                new CallExpression(aggregation.getCall().getDisplayName(), aggregation.getCall().getFunctionHandle(), aggregation.getCall().getType(), argumentWithoutPercentile.build()),
+                aggregation.getFilter(),
+                aggregation.getOrderBy(),
+                aggregation.isDistinct(),
+                aggregation.getMask());
+    }
+
+    @Override
+    public Result apply(AggregationNode aggregation, Captures captures, Context context)
+    {
+        ImmutableMap.Builder<VariableReferenceExpression, AggregationNode.Aggregation> aggregations = ImmutableMap.builder();
+
+        // VariableReferenceExpression is the output of aggregation, RowExpression is the percentile argument
+        Map<AggregationNode.Aggregation, Map<VariableReferenceExpression, RowExpression>> commonApprox = new HashMap<>();
+
+        // Collect candidate approx_percentile calls which can be combined
+        for (Map.Entry<VariableReferenceExpression, AggregationNode.Aggregation> entry : aggregation.getAggregations().entrySet()) {
+            AggregationNode.Aggregation curAggregation = entry.getValue();
+            // Skip aggregation which already takes array as percentile arguments
+            if (curAggregation.getCall().getDisplayName().equals(APPROX_PERCENTILE) && curAggregation.getCall().getArguments().size() >= 2 && !(curAggregation.getCall().getType() instanceof ArrayType)) {
+                int percentilePosition = getPercentilePosition(curAggregation.getCall().getFunctionHandle());
```

Comment:
Here also use function handle

Reply:
Discussed offline, it's already resolved.